exonum-java-binding/core/rust/src/proxy/runtime.rs

```diff
@@ -0,0 +1,596 @@
+            let view_handle = to_handle(View::from_ref_fork(context.fork));
+            let pub_key = JObject::from(env.byte_array_from_slice(&tx.0.to_bytes())?);
+            let hash = JObject::from(env.byte_array_from_slice(&tx.1.to_bytes())?);
+
+            env.call_method_unchecked(
+                self.runtime_adapter.as_obj(),
+                runtime_adapter::execute_tx_id(),
+                JavaType::from_str("V").unwrap(),
+                &[
+                    JValue::from(service_id),
+                    JValue::from(tx_id),
+                    JValue::from(args),
+                    JValue::from(view_handle),
+                    JValue::from(pub_key),
+                    JValue::from(hash),
+                ],
+            )?;
+            Ok(())
+        }))
+    }
+
+    fn state_hashes(&self, snapshot: &Snapshot) -> StateHashAggregator {
+        let bytes = unwrap_jni(self.exec.with_attached(|env| {
+            let view_handle = to_handle(View::from_ref_snapshot(snapshot));
+            let java_runtime_hashes = panic_on_exception(
+                env,
+                env.call_method_unchecked(
+                    self.runtime_adapter.as_obj(),
+                    runtime_adapter::state_hashes_id(),
+                    JavaType::from_str("[B").unwrap(),
+                    &[JValue::from(view_handle)],
+                ),
+            );
+            let byte_array = java_runtime_hashes.l()?.into_inner();
+            let data = env.convert_byte_array(byte_array)?;
+
+            Ok(data)
+        }));
+
+        ServiceRuntimeStateHashes::from_bytes(bytes.into()).unwrap().into()
+    }
+
+    fn before_commit(&self, _dispatcher: &Dispatcher, _fork: &mut Fork) {
+        // TODO: is not supported by ServiceRuntimeAdapter
+    }
+
+    fn after_commit(
+        &self,
+        dispatcher: &DispatcherSender,
+        snapshot: &Snapshot,
+        service_keypair: &(PublicKey, SecretKey),
+        tx_sender: &ApiSender
+    ) {
+        let context = AfterCommitContext::new(dispatcher, snapshot, service_keypair, tx_sender);
+
+        unwrap_jni(self.exec.with_attached(|env| {
+            let view_handle = context.view_handle();
+            let validator_id = context.validator_id_or(-1);
+            let height = context.height().0 as i64;
+
+            panic_on_exception(
+                env,
+                env.call_method_unchecked(
+                    self.runtime_adapter.as_obj(),
+                    runtime_adapter::after_commit_id(),
+                    JavaType::from_str("V").unwrap(),
+                    &[
+                        JValue::from(view_handle),
+                        JValue::from(validator_id),
+                        JValue::from(height),
+                    ],
+                ),
+            );
+            Ok(())
+        }));
+    }
+
+    fn api_endpoints(&self, context: &ApiContext) -> Vec<(String, ServiceApiBuilder)> {
```

Comment:
This method won't work, as there is no way to attach Java endpoints to the actix-web server of Core. This method is removed in latest Core revision.

Reply:
Sorry, my fault, the method is not removed, but can be used only by Rust runtime